### ISSUE TITLE
torch compile + replicate, compute and communication not overlap

### ISSUE THREAD

**Issue by peaceorwell** [pytorch/pytorch#169461]
### 🐛 Describe the bug

When I use a combination of composable.replicate and torch.compile, I observe that all backward allreduce operations are executed only after the entire backward pass computation is complete.

This behavior prevents the overlap of computation and communication, which is typically achieved in DDP (DistributedDataParallel) + torch.compile by inserting a graph break during the backward pass (e.g., after the gradient calculation for a specific layer).

I am looking for any potential workarounds or suggested methods to enable computation/communication overlap when using composable.replicate with torch.compile.

```
import os
import time
import torch
import torch.distributed as dist
import torch.nn as nn

from torch.distributed._composable.replicate import replicate

from torch.profiler import profile, record_function, ProfilerActivity

def setup():
    rank = int(os.environ["RANK"])
    local_rank = int(os.environ["LOCAL_RANK"])
    world_size = int(os.environ["WORLD_SIZE"])

    dist.init_process_group("nccl")
    torch.cuda.set_device(local_rank)
    
    return rank, local_rank

def cleanup():
    dist.destroy_process_group()

class ToyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.layers = nn.Sequential(
            nn.Linear(2048, 4096),
            nn.ReLU(),
            nn.Linear(4096, 4096),
            nn.ReLU(),
            nn.Linear(4096, 4096),
            nn.ReLU(),
            nn.Linear(4096, 2048),
        )

    def forward(self, x):
        return self.layers(x)

def main():
    rank, local_rank = setup()

    model = ToyModel().to(local_rank)

    replicate(
        model, 
        device_ids=[local_rank], 
        bucket_cap_mb=25
    )

    opt_model = torch.compile(model, backend="inductor")

    optimizer = torch.optim.SGD(opt_model.parameters(), lr=0.01)
    loss_fn = nn.MSELoss()

    input_tensor = torch.randn(32, 2048).to(local_rank)
    target_tensor = torch.randn(32, 2048).to(local_rank)

    log_dir = './profiler_logs_composable'
    
    with profile(
        activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
        schedule=torch.profiler.schedule(
            wait=5,
            warmup=2,
            active=3,
            repeat=1
        ),
        on_trace_ready=torch.profiler.tensorboard_trace_handler(log_dir),
        record_shapes=True,
        profile_memory=True,
        with_stack=True
    ) as prof:
        
        for step in range(15):
            step_start = time.time()
            
            with record_function("model_training_step"):
                optimizer.zero_grad()
                
                output = opt_model(input_tensor)
                loss = loss_fn(output, target_tensor)
                loss.backward()
                optimizer.step()
            
            torch.cuda.synchronize()
            step_end = time.time()

            if rank == 0:
                print(f"Step {step}: Loss={loss.item():.4f}, Time={step_end - step_start:.4f}s")
            
            prof.step()

    cleanup()

if __name__ == "__main__":
    main()
```

![Image](https://github.com/user-attachments/assets/0af08b18-893b-4cff-b8fe-b23cb10fb4be)

### Versions

PyTorch version: 2.10.0a0+b558c986e8.nv25.11
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.39

Python version: 3.12.3 (main, Aug 14 2025, 17:47:21) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-150-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-PCIE-40GB
GPU 1: NVIDIA A100-PCIE-40GB

Nvidia driver version: 570.133.20
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.15.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.15.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          112
On-line CPU(s) list:             0-111
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              2
Core(s) per socket:              28
Socket(s):                       2
Stepping:                        6
Frequency boost:                 enabled
CPU(s) scaling MHz:              28%
CPU max MHz:                     3100.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4000.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       2.6 MiB (56 instances)
L1i cache:                       1.8 MiB (56 instances)
L2 cache:                        70 MiB (56 instances)
L3 cache:                        84 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-27,56-83
NUMA node1 CPU(s):               28-55,84-111
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] intel-openmp==2021.4.0
[pip3] mkl==2021.1.1
[pip3] mkl-devel==2021.1.1
[pip3] mkl-include==2021.1.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.1.0
[pip3] nvidia-cudnn-frontend==1.15.0
[pip3] nvtx==0.2.13
[pip3] onnx==1.18.0
[pip3] onnx-ir==0.1.12
[pip3] onnxscript==0.5.6
[pip3] optree==0.17.0
[pip3] pytorch-triton==3.5.0+gitde3506d2
[pip3] tbb==2021.13.1
[pip3] torch==2.10.0a0+b558c986e8.nv25.11
[pip3] torch_tensorrt==2.10.0a0
[pip3] torchao==0.14.0+git
[pip3] torchdata==0.11.0
[pip3] torchprofile==0.0.4
[pip3] torchtitan==0.1.0
[pip3] torchvision==0.25.0a0+7a13ad0f
[pip3] triton_kernels==1.0.0+nv25.11
[conda] Could not collect

@fmassa @eellison @bdhirsh @zou3519 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @chauhang @penguinwu

**Comment by HarryZJ** [comment]
I found the replicate interface can't overlap communication and computation like DDP does  is likely  because DistributedDataParallel._active_ddp_module isn't set, making DDPOptimizer ineffective. I hacked the `_ReplicateState`  class code (3 lines below，marked “for debug”，) to enable DDPOptimizer and get overlapping to work—could this be a bug in replicate?

``` 
def forward_pre_hook(
        self, module: nn.Module, args: tuple[Any, ...], kwargs: dict[str, Any]
    ) -> Any:
        if self._init_args or self._init_kwargs:
            self.lazy_init()
        self._ddp.require_backward_grad_sync = not self._no_sync
        DistributedDataParallel._active_ddp_module = module  # for debug 
        module.bucket_bytes_cap = self._ddp.bucket_bytes_cap # for debug
        return self._ddp._pre_forward(*args, **kwargs)

def forward_post_hook(
        self,
        module: nn.Module,
        input: tuple[torch.Tensor],
        output: torch.Tensor,
    ) -> torch.Tensor:
        DistributedDataParallel._active_ddp_module = None # for debug
        return self._ddp._post_forward(output)
```

<img width="1041" height="550" alt="Image" src="https://github.com/user-attachments/assets/313760b0-da4b-4ac7-a81a-ef1aa380a624" />
